Added UnitTest & fix for TestInfoPrivateAddressDisclosure

schoeth opened the pull request:
Changes, as mentioned in #2834:
- prefix AWS-hostnames (ip-)
- Fixed a typo in a regex
- updated the alert-description and help page
  Added a UnitTest acc. to #2790

This closes zaproxy/zaproxy#2834 and zaproxy/zaproxy#2790

thc202 commented on src/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosure.java:
```diff
@@ -1,153 +1,159 @@
-/*
```
Line ending changes?

schoeth replied:
I have no idea what's wrong here. 
The Eclipse-diff shows me just the 16 edited lines.
Copied the files from master and  issue-2834... to another folder and a comparison with Notepad++ also shows just the really changed lines.
Both files have CRLFs, and are UTF-8 (which shouldn't matter here)
But interesting, that this is the only file with that problem in GitHub-diff

thc202 replied:
In seems master is using CRLF but the file in this branch is using just LF.

schoeth replied:
As written, both files on HDD have CRLF.
Will try it with to converting forth and back...

schoeth replied:
OK, converting to just LF didn't help, convert back now...

thc202 replied:
Might be because of:
https://www.kernel.org/pub/software/scm/git/docs/gitattributes.html#_end_of_line_conversion

schoeth replied:
Changed back, but still the same here - and my local Eclipse & Notebook++ - diffs show still just mark the really edited lines.

thc202 replied:
Something is changing the line endings then (git? EGit?), worth checking the git configs in your local repo.

schoeth commented:
Just added tests to the unitTestCase:
- that just the first finding is in  the evidence,
- a "dummy-test" for the ignred `scanHttpRequestSend`
  (if code coverage should get introduced in the future)
- good idea from adioss in PR#2786: also validate the other Alert-values

kingthorin commented on src/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosure.java:
```diff
@@ -36,29 +36,35 @@
 	
     private static final String REGULAR_IP_OCTET = "\\b(25[0-5]|2[0-4][0-9]|1?[0-9]{1,2})";
-    private static final String REGULAR_PORTS = "\\b(6553[0-5]|65[0-5][0-2][0-9]|6[0-4][0-9]{4}|[0-5]?[0-9]{0,4})";
+    private static final String REGULAR_PORTS = "\\b(6553[0-5]|65[0-5][0-2][0-9]|6[0-4][0-9]{3}|[0-5]?[0-9]{0,4})";
```
@thc202 @schoeth I think I've identified another mistake in this RegEx. I believe it should also end with an word boundry marker `\\b` otherwise `:75000` matches on the first 4 characters (_7500_).

Original: https://regex101.com/r/uO1rI2/2
Proposed fix: https://regex101.com/r/uO1rI2/3
Note: Regex101 syntax is slightly different, due to escaping requirements.

schoeth replied:
hehe... remember my comments in #2834?  (i know i'm an old boring '1stplanner-than-doer' ;)
I'll check that, and also the resulting effect of e.g. `10.0.0.0:65536` -> `10.0.0.0:` in evidence, and commit a next proposal.
BTW, thanks for the regex101-link, didn't know it!

schoeth replied:
@kingthorin yes, regex101 has no exact java-syntax, but they seem to provide 3 other variants, which is interesting to me.
But don't worry, the solution for the above issue is already known, just need time to re-code the tests.

schoeth commented:
i still fight my eGit, my last commit was:
Changed regex to catch all ports 0-65535 without any exceptions, ...
added boundaries to IPs & AWS-HN,
changed unit tests accordingly,
impl. Alert.otherInfo to a unit test

**Please check**, if this functionality is OK:
- Boundaries, e.g. unit test line 117,118
- chatches ALL ports 0-65535, see unit test lines from 135

kingthorin commented:
Looks good to me

thc202 commented:
@schoeth could you add the ZAP comment to the scanner (it's a Paros class)?

thc202 commented on src/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosure.java:
```diff
@@ -36,47 +36,36 @@
+     */
+    private static final String MESSAGE_PREFIX = "pscanrules.testinfoprivateaddressdisclosure.";
+
+    private static final String REGULAR_IP_OCTET = "\\b(25[0-5]|2[0-4][0-9]|1?[0-9]{1,2})\\b";
```
For the record there are couple of word boundaries that could be dropped (e.g. drop from the octet and add to the begin/end of the addresses; drop from the beginning of the port) and some unnecessary escaped characters.

thc202 commented on src/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosure.java:
```diff
@@ -148,6 +137,5 @@ private int getRisk() {
     @Override
     public void setParent(PassiveScanThread parent) {
         this.parent = parent;
-
```
This could be reverted.

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,334 @@
+			HttpMessage msg = createHttpMessage(candidate);
+			rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+			assertThat(alertsRaised.size(), equalTo(i + 1));
```
It would be good to include the `candidate` as reason when asserting the sizes, it's easier to know which IP is failing (if that happens someday).

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,334 @@
+				HttpMessage msg = createHttpMessage(candidate);
+				rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+				
+				int alertNoIdx = ci * 2 + ii;
```
The list of alerts raised could be cleared prior scanning, reduces the memory used and the index would be always the same.

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,334 @@
+
+	@Test
+	public void alertsIfPrivIpAndAddsPortToEvidence() throws HttpMalformedHeaderException {
+		// ip and aws-hostname which get caoncatenated with the ports as candidates
```
There's a typo in "concatenated" (same in other test).

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,334 @@
+	}
+
+	@Test
+	public void TestOfScanHttpRequestSend() throws HttpMalformedHeaderException {
```
The name should start with lower case.

kingthorin replied:
First char

thc202 commented on src/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosure.java:
```diff
@@ -36,47 +36,36 @@
-            + "172\\-" + "\\b(3[01]|2[0-9]|1[6-9])\\-" + REGULAR_IP_OCTET + "\\-" + REGULAR_IP_OCTET + "|"
-            + "192\\-168\\-" + REGULAR_IP_OCTET + "\\-" + REGULAR_IP_OCTET             
+            + "\\bip-10-(" + REGULAR_IP_OCTET + "-){2}" + REGULAR_IP_OCTET + "\\b|"
+            + "\\bip-172-" + "(3[01]|2[0-9]|1[6-9])-" + REGULAR_IP_OCTET + "-" + REGULAR_IP_OCTET + "|"
```
It's missing the word boundary at the end (before `|`).

thc202 commented on src/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosure.java:
```diff
@@ -36,47 +36,36 @@
-    //"(10\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}|172\\.\\d{2,2}\\.\\d{1,3}\\.\\d{1,3}|192\\.168\\.\\d{1,3}\\.\\d{1,3})", PATTERN_PARAM);
-    
+            // find regular ports (0-65535)
+            + "(:\\b(0|[1-9]\\d{0,3}|[1-5]\\d{4}|6[0-4]\\d{3}|65([0-4]\\d{2}|5[0-2]\\d|53[0-5]))\\b)?",
```
The word boundary after the colon could be removed?

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,337 @@
+				HttpMessage msg = createHttpMessage(candidate);
+				rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+				
+				assertThat(alertsRaised.size(), equalTo(1));
```
It could be added the candidate, to help in case of failure.

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,337 @@
+				HttpMessage msg = createHttpMessage(candidate);
+				rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+				
+				assertThat(alertsRaised.size(), equalTo(1));
```
It could be added the candidate, to help in case of failure.

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,337 @@
+			HttpMessage msg = createHttpMessage(candidate);
+			rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+	
+			assertThat(alertsRaised.size(), equalTo(i + 1));
```
It could be added the candidate, to help in case of failure.

thc202 commented on test/org/zaproxy/zap/extension/pscanrules/TestInfoPrivateAddressDisclosureUnitTest.java:
```diff
@@ -0,0 +1,337 @@
+			HttpMessage msg = createHttpMessage(data[i][0]);
+			rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+			assertThat(alertsRaised.size(), equalTo(i + 1));
```
It could be added the candidate, to help in case of failure.

thc202 commented:
Could you update the "changes" in the ZapAddOn.xml file of the add-on, to mention that it fixes the false positives?
The file was changed upstream so better to rebase before changing it.

thc202 commented:
Thank you!
